### ISSUE TITLE
Constant CPU TImeouts or DML Limits hit on a small amount of rollups

### ISSUE THREAD

**Issue by tony-jones-tanium** [jamessimone/apex-rollup#404]
Hi, 

We are having an issue where a single rollup is erroring out, due to either CPU Timeout or Too many DML rows: 10001. I have adjusted the Query limits and sizes and still running into the error when there are about 10 records to update.

The Child Object Where Clause, should be limiting the records that are being looked at as well. 

![ARR Rollup](https://user-images.githubusercontent.com/103075625/214112279-f7534b8d-ad19-45d8-b9f9-f3c530d08e13.png)
![Rollup Control](https://user-images.githubusercontent.com/103075625/214112297-1ae20b71-b87b-4399-b517-ac0fbb0127fb.png)

**Comment by jamessimone** [comment]
@tony-jones-tanium It'd be great to be able to see a log or at least the extended stacktrace of where this is happening. That would help me to understand the entry point (flow, apex) as well as where this error is being generated. There's nothing obviously wrong with what you've included, so I'm wondering if there aren't downstream updates that are causing this error to occur

**Comment by tony-jones-tanium** [comment]
Here is an error log from each. At this point, Im just trying to manually execute the rollups, but it errored out from a scheduled job on both Saturday and Sunday. This is also a sandbox environment. 
![CPU Error](https://user-images.githubusercontent.com/103075625/214116776-8d750694-60cd-4750-a112-841e42235ebe.png)
![DML Error](https://user-images.githubusercontent.com/103075625/214116798-4ac128e5-1149-4cc9-93a6-2817df1c8411.png)

**Comment by tony-jones-tanium** [comment]
Depending on what I change the Max Query total to (25,000 or 10,000), I get one or the other, in terms of errors: 
![image](https://user-images.githubusercontent.com/103075625/214119177-df4acafb-981f-42b4-a9c7-e5ef8134c135.png)

**Comment by jamessimone** [comment]
OK there's a few things I'm wondering if we can't play with here:

* have you tried lowering the `Max Parent Rows Updated At Once` field on your Rollup Control? It sounds like there are downstream automations that are making it so that 1000 records (the default) in this setup are each updating 10 or more records, leading to the first issue you were experiencing
* have you tried to `Max Lookup Rows Before Batching` to something lower than the 3000 default? 

Additionally - do you also have rollups set up on your `Account` object? In looking at the stacktrace you included, it seems like a full recalc run is being performed - that error is coming from `RollupParentResetProcessor`, and that class is typically only involved in full recalc runs. Additionally, that error is telling me something else - either in the rollup that you sent me, or in one on `Account` (if the stacktrace is actually coming from a downstream rollup), you have a lot of records that meet the _negative case_ for your rollup. 

In other words, what the `RollupParentResetProcessor` tries to do is find records that have values for the Rollup Field On Parent Object but _were not_ included in the scheduled full recalc. Especially if, as I alluded to above, you have one-to-many automation on that parent object that could lead to too many updates being made, my first bullet point may help address this problem fully by chunking the parent-level updates into more manageable levels.

**Comment by tony-jones-tanium** [comment]
Hey @jamessimone! Thanks so much for the tips. 

- I tried lowing the Max Parent Rows and Max Lookup rows, and continued to manually use the Recalculate Rollup (using CMDT) and still got the CPU Timeouts: 

![Screenshot 2023-01-23 at 1 49 49 PM](https://user-images.githubusercontent.com/103075625/214158098-d5519c00-6b64-43a5-8cc3-74ad538d2c78.png)
![Screenshot 2023-01-23 at 1 50 09 PM](https://user-images.githubusercontent.com/103075625/214158118-a4c88f34-c8c1-4021-b67e-24a3d08ea5c8.png)

We have a scheduled flow that has 8 different Apex Rollups that runs at night, but I have those separated out by pauses for 10 minutes. In this case, Im just trying to manually Recalc, so that I know the Accounts have been updated. That being said, we do have a good amount of Opportunities (Child records) that will NOT meet the WHERE Clause criteria right now in this sandbox. So maybe thats the reason?

**Comment by jamessimone** [comment]
I've been thinking about putting the parent reset run behind a configuration flag, as it's been a pain point for some other users as well. This seems like a good reason to do that - I should be able to get a release for that out there today.

**Comment by tony-jones-tanium** [comment]
Thanks @jamessimone! Im happy to test once that's release, as Im working on this all week. 

 I just tried a different approach. Changing the time of my scheduled flow to now, and it seems like all of the other Apex Rollups that are triggered from the scheduled flow, are updating, at least. 
 
 Its just odd that its not updating the single record that meets the criteria. 
 
 
<img width="1119" alt="Screenshot 2023-01-23 at 2 40 50 PM" src="https://user-images.githubusercontent.com/103075625/214171488-f4d92294-5cb2-43e0-a9e3-aa851c1a454a.png">
<img width="1373" alt="Screenshot 2023-01-23 at 2 39 37 PM" src="https://user-images.githubusercontent.com/103075625/214171503-d18fb7f5-1517-4e86-ad2c-e9a59a978ee6.png">

**Comment by jamessimone** [comment]
@tony-jones-tanium turns out I added this feature a while back and had forgotten - apologies for the oversight. You should be able to remove those CPU timeouts by also toggling `Should Skip Resetting Parent Fields` to false on your default Rollup Control record.

**Comment by tony-jones-tanium** [comment]
Hey @jamessimone! I just added that field to our Rollup Control layout, marked it as TRUE, tried to run the manual Recalculation of the Rollup again and got the same CPU timeout, when there is a single record that needs to be updated still. 
<img width="1092" alt="Screenshot 2023-01-24 at 8 43 07 AM" src="https://user-images.githubusercontent.com/103075625/214354966-b0a2f37a-4461-4fc8-a0d7-aaceae62746f.png">
<img width="1328" alt="Screenshot 2023-01-24 at 8 43 30 AM" src="https://user-images.githubusercontent.com/103075625/214354986-3b981839-bae4-45dd-be79-8be292f5f2f2.png">
<img width="1124" alt="Screenshot 2023-01-24 at 8 43 49 AM" src="https://user-images.githubusercontent.com/103075625/214355005-9c60dc89-a12f-4e40-9576-d868235bde1c.png">

**Comment by jamessimone** [comment]
🤦‍♂️ I'll add that field to the layout in the next release. I'll take another look at the parent reset processor - it shouldn't be running at all based on you having updated that flag, but clearly there's still some bug here I haven't yet considered.

**Comment by jamessimone** [comment]
@tony-jones-tanium I will probably move to release the changes from discovery here tomorrow. Something to note - it looks like your `Batch Chunk Size` is still set to the maximum value of 2000. I would recommend bringing that down by quite a bit, as that is going to be the single biggest contributor to CPU time on a per-batch basis. The fixes I'm planning to release tomorrow will almost certainly fix that error you were seeing with the `RollupParentResetProcessor`.

**Comment by jamessimone** [comment]
@tony-jones-tanium the latest release should address the issues you were having with the parent reset processor - and I suspect the other issues will be solved with an update to the Batch Chunk Size, as noted above. Please keep me informed!

**Comment by jamessimone** [comment]
Are you saying that the parent reset processor is still running even after updating with the `Should Skip Resetting Parent Fields` flag set to true? That should not be the case. 

I still think the primary issue here is that the `Batch Chunk Size` variable needs to be lowered.

**Comment by tony-jones-tanium** [comment]
Hey @jamessimone! Sorry for the delayed response, today got away from me! I actually was able to get both the manual recalc AND our scheduled flow to work early this morning with the Batch Chuck Size reduction, and the installation of the newest version. I still need to update in our Production environment next week during our next scheduled deployment to confirm it will work there.

**Comment by jamessimone** [comment]
Not to worry - apologies for the double message. the CPU clock on this particular computer is messed up and it zoomed me to an earlier part of our conversation. Glad to hear it, though! I'll await your confirmation before closing this issue as resolved.

**Comment by jamessimone** [comment]
@tony-jones-tanium I wanted to follow up with you about this one - is all well on your end?

**Comment by tony-jones-tanium** [comment]
Hey @jamessimone! Just wanted to report back that everything went smoothly with our scheduled Apex Rollups last evening. Thats so much for your help on this, as this is a big win for our team!

**Comment by jamessimone** [comment]
Awesome, happy to hear it!
